### DISCUSSION TITLE
Passing calculated values to random_int does not work as expected

### DISCUSSION THREAD

**Discussion by sagikazarmark** [redpanda-data/connect#2110] [General]
I'm trying to pick a random element from a list. Here is what I tried so far:

```bloblang
let regions = ["us-east-1", "us-west-1", "us-east-2", "us-west-2"]

let region = $regions.index(random_int(max: $regions.length()))
```

However, this always returns the last item in the list. Checking the output of the call to `random_int` reveals that it always outputs 4.

Changing `$regions.length()` to a static number works as expected.

Changing the code to the following also works fine:

```bloblang
let regions = ["us-east-1", "us-west-1", "us-east-2", "us-west-2"]

let region = $regions.index(random_int() % $regions.length())
```

Not sure if this is a bug or a known limitations. I also wasn't sure how to search the issue tracker for potentially existing issues, my attempts didn't reveal any.

**Comment by mihaitodor** [comment]
Hey @sagikazarmark that's because the `random_int()` function has a `seed` parameter which is set to `0` by default (docs [here](https://www.benthos.dev/docs/guides/bloblang/functions#random_int)). Try `random_int(seed:timestamp_unix_nano(), max: $regions.length())`.

PS: Converting this into a discussion as mentioned in #2026.

**Reply by sagikazarmark** [reply]
@mihaitodor appreciate the quick answer.

Can you tell me why Whis works?

```
let region = $regions.index(random_int(max: 4))
```

It always returns a random number. Also when no params are passed to the function. It only fails in the first case.

**Reply by mihaitodor** [reply]
Indeed, you're right! I should've thought more about this before answering. Glad @Jeffail chimed in ❤️

**Comment by Jeffail** [comment]
This is actually kind of a bug, or at least a limitation of bloblang. The problem with your mapping is that the argument for `max` is dynamic as the evaluation of the variable `$regions` is redone for each invocation. Since the arguments for the `random_int` function are dynamic it means the random number generator is recreated for each invocation, so since the seed is static the generator is always created brand new for each mapping execution and that's why you get the same value multiple times.

We could expand the static evaluation of mappings in order to detect static variables, which in the example you posted would fix things as the arguments would no longer be dynamic. However, that wouldn't solve the greater issue of when those arguments aren't capable of being statically evaluated.

The real bug here is that the random number generator is being recreated (and therefore reset) and it's not obvious to you as a mapping author or the config executor why that is happening, or even that it's happening at all.

We could potentially maintain a cache of generators that have been created, since by their nature they're meant to be random I think it'd be reasonable to have a shared cache. As a short term work-around you can implement a max the old fashioned way with `random_int() % $regions.length()`. Either way the current behaviour is a bug (at the very least in documentation) so I'm going to reopen the original issue.

**Reply by sagikazarmark** [reply]
Thanks @Jeffail ! I'm comfortable with the old-fashioned solution, I just wanted to raise this with you in case it was a bug (and apparently it is). But I can live with the current solution.

Thanks again for the detailed reply!

**Comment by Jeffail** [comment]
Can't reopen it now that it's been converted so I've opened https://github.com/benthosdev/benthos/issues/2112
